homeassistant/components/qbus/qbus.py

```diff
@@ -0,0 +1,78 @@
+"""Qbus classes."""
+
+import asyncio
+import logging
+from typing import Final
+
+from qbusmqttapi.discovery import QbusDiscovery
+from qbusmqttapi.factory import QbusMqttTopicFactory
+
+from homeassistant.components.mqtt import async_wait_for_mqtt_client, client as mqtt
+from homeassistant.core import HomeAssistant
+
+from .const import DATA_QBUS_CONFIG, DATA_QBUS_CONFIG_EVENT, DOMAIN
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class QbusConfigContainer:
+    """Helper to handle the Qbus config."""
+
+    _WAIT_TIMEOUT: Final[int] = 30
+    _topic_factory = QbusMqttTopicFactory()
+
+    @staticmethod
+    async def async_get_or_request_config(hass: HomeAssistant) -> QbusDiscovery | None:
+        """Get or request Qbus config."""
+        hass.data.setdefault(DOMAIN, {})
+        domain_data: dict = hass.data[DOMAIN]
```

Comment:
instead of having static methods and storing the data on hass.data, can we make QbusConfigContainer an instance variable of the Coordingator? And maybe call it simply QbusConfig

Reply:
The thing is that the data published to the config topic contains the configuration for all devices at once, while the coordinator is a per device class. My goal was to avoid requesting the same config multiple times (in case of multiple devices), so I chose to store the config somewhere globally.

Is there a way to do this without hass.data and/or static methods?

Let me also take this opportunity to thank you for the effort you put in this! I can imagine it’s difficult to follow the code while not being able to see what is happening or what it is supposed to do. And it’s good to be challenged by your questions too!

Reply:
We could change the `QbusConfigContainer` class to have normal instance methods, and store the `QbusConfigContainer` instance on hass.data. I was just made aware that wled does also has a global coordinator and one per config entry (on `entry.runtime_data): https://github.com/Qbus-iot/core/blob/64bb835ae2ec5f17e6735ca3675513665aca164f/homeassistant/components/wled/__init__.py#L38
You can also use HassKey for it to be typed.

Maybe you can also move the config topic handling to `QbusConfigContainer`? Otherwise multiple coordinators will be calling `QbusConfigContainer.store_config()` with the same data.

Reply:
We could change the QbusConfigContainer class to have normal instance methods, and store the QbusConfigContainer instance on hass.data. I was just made aware that wled does also has a global coordinator and one per config entry (on `entry.runtime_data): Qbus-iot/core@64bb835/homeassistant/components/wled/__init__.py#L38
You can also use HassKey for it to be typed.

Maybe you can also move the config topic handling to QbusConfigContainer? Otherwise multiple coordinators will be calling QbusConfigContainer.store_config() with the same data.

Reply:
`QbusConfigContainer` has been refactored into `QbusConfigCoordinator`. Also following approach of the wled integration for setting up and using the `QbusConfigCoordinator`.